lib/update/cluster/builder.go

```diff
@@ -0,0 +1,672 @@
+}
+
+// migration constructs a migration phase based on the plan params.
+//
+// If there are no migrations to perform, returns nil.
+func (r phaseBuilder) migration(leadMaster storage.Server, p planConfig) *update.Phase {
+	root := update.RootPhase(update.Phase{
+		ID:          "migration",
+		Description: "Perform system database migration",
+	})
+
+	var subphases []update.Phase
+
+	// do we need to migrate links to trusted clusters?
+	if len(p.links) != 0 && len(p.trustedClusters) == 0 {
+		subphases = append(subphases, update.Phase{
+			ID:          root.ChildLiteral("links"),
+			Description: "Migrate remote Ops Center links to trusted clusters",
+			Executor:    migrateLinks,
+		})
+	}
+
+	// Update / reset the labels during upgrade
+	subphases = append(subphases, update.Phase{
+		ID:          root.ChildLiteral("labels"),
+		Description: "Update node labels",
+		Executor:    updateLabels,
+	})
+
+	// migrate roles
+	if libphase.NeedMigrateRoles(p.roles) {
+		subphases = append(subphases, update.Phase{
+			ID:          root.ChildLiteral("roles"),
+			Description: "Migrate cluster roles to a new format",
+			Executor:    migrateRoles,
+			Data: &storage.OperationPhaseData{
+				ExecServer: &leadMaster,
+			},
+		})
+	}
+
+	// no migrations needed
+	if len(subphases) == 0 {
+		return nil
+	}
+
+	root.AddParallel(subphases...)
+	return &root
+}
+
+// Only applicable for 5.3.0 -> 5.3.2
+// We need to update the CoreDNS app before doing rolling restarts, because the new planet will not have embedded
+// coredns, and will instead point to the kube-dns service on startup. Updating the app will deploy coredns as pods.
+// TODO(knisbet) remove when 5.3.2 is no longer supported as an upgrade path
+func (r phaseBuilder) earlyDNSApp(locator loc.Locator) *update.Phase {
+	phase := update.Phase{
+		ID:       locator.Name,
+		Executor: updateApp,
+		Description: fmt.Sprintf(
+			"Update system application %q to %v", locator.Name, locator.Version),
+		Data: &storage.OperationPhaseData{
+			Package: &locator,
+		},
+	}
+	return &phase
+}
+
+// config returns phase that pulls system configuration on provided nodes
+func (r phaseBuilder) config(nodes []storage.Server) *update.Phase {
+	root := update.RootPhase(update.Phase{
+		ID:          "config",
+		Description: "Update system configuration on nodes",
+	})
+	for i, node := range nodes {
+		root.AddParallel(update.Phase{
+			ID:       root.ChildLiteral(node.Hostname),
+			Executor: config,
+			Description: fmt.Sprintf("Update system configuration on node %q",
+				node.Hostname),
+			Data: &storage.OperationPhaseData{
+				Server: &nodes[i],
+			},
+		})
+	}
+	return &root
+}
+
+func (r phaseBuilder) runtime(updates []loc.Locator) *update.Phase {
+	root := update.RootPhase(update.Phase{
+		ID:          "runtime",
+		Description: "Update application runtime",
+	})
+	sort.Slice(updates, func(i, j int) bool {
+		// Push RBAC package update to front
+		return updates[i].Name == constants.BootstrapConfigPackage
```

Comment:
Instead of searching for RBAC app in the builder step and threading a boolean parameter - this does away with this and simply pushes the RBAC update to the front and makes all apps depend upon each other. As long as we don't support parallel execution - it is as good as before.